lib/api/multiObjectDelete.js

```diff
@@ -0,0 +1,307 @@
+                            return next(errors.InternalError);
+                        }
+                        const key = result.arn.slice(slashIndex + 1);
+                        if (result.isAllowed) {
+                            inPlay.push(key);
+                        } else {
+                            errorResults.push({
+                                key,
+                                error: errors.AccessDenied,
+                            });
+                        }
+                    }
+                    return next(null, quietSetting);
+                });
+        },
+        function checkBucketMetadata(quietSetting, next) {
+            // if no objects in play, no need to check ACLs,
+            // just move on
+            if (inPlay.length === 0) {
+                return next(null, quietSetting);
+            }
+            return metadata.getBucket(bucketName, log, (err, bucketMD) => {
+                if (err) {
+                    log.trace('error retrieving bucket metadata');
+                    return next(err);
+                }
+                // check whether bucket has transient or deleted flag
+                if (bucketShield(bucketMD, 'objectDelete')) {
+                    return next(errors.NoSuchBucket);
+                }
+                if (!isBucketAuthorized(bucketMD, 'objectDelete',
+                    canonicalID)) {
+                    log.trace('access denied due to bucket acl\'s');
+                    // if access denied at the bucket level, no access for
+                    // any of the objects so all results will be error results
+                    inPlay.forEach(key => {
+                        errorResults.push({
+                            key,
+                            error: errors.AccessDenied,
+                        });
+                    });
+                    // by setting inPlay length to 0 it empties out the
+                    // array so async.forEachLimit below will not actually
+                    // make any calls to metadata or data but will continue on
+                    // to the next step to build xml
+                    inPlay.length = 0;
+                }
+                return next(null, quietSetting);
+            });
+        },
+        function getObjMetadataAndDelete(quietSetting, next) {
+            // for obj deletes, no need to check acl's at object level
+            // (authority is at the bucket level for obj deletes)
+
+            // doing 5 requests at a time. note that services.deleteObject
+            // will do 5 parallel requests to data backend to delete parts
+            return async.forEachLimit(inPlay, 5, (key, moveOn) => {
+                metadata.getObjectMD(bucketName, key, log, (err, objMD) => {
+                    // if general error from metadata return error
+                    if (err && !err.NoSuchKey) {
+                        return next(err);
```

Comment:
This means that in case of an error during the process, we do not continue with the process at all, and do not send back which elements we deleted/ errored out, is that something desired?